subcommands/targets/offline-update.go

```diff
@@ -0,0 +1,286 @@
+package targets
+
+import (
+	"archive/tar"
+	"compress/bzip2"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/foundriesio/fioctl/client"
+	"github.com/foundriesio/fioctl/subcommands"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"io"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	factory string
+	prod    bool
+	tag     string
+)
+
+type (
+	targetInfo struct {
+		version       int
+		ostreeVersion int
+		hardwareID    string
+		buildTag      string
+	}
+)
+
+func init() {
+	offlineUpdateCmd := &cobra.Command{
+		Use:   "offline-update <target-name> <dst> --tag <tag> [--prod]",
+		Short: "Download Target content for an offline update",
+		Run:   doOfflineUpdate,
+		Args:  cobra.ExactArgs(2),
+		Example: `
+	# Download update content of the production target #1451 tagged by "release-01" for "intel-corei7-64" hardware type
+	fioctl targets offline-update intel-corei7-64-lmp-1451 /mnt/flash-drive/offline-update-content --tag release-01 --prod
+
+	# Download update content of the CI target #1451 tagged by "devel" for "raspberrypi4-64" hardware type
+	fioctl targets raspberrypi4-64-lmp-1448 /mnt/flash-drive/offline-update-content --tag devel
+
+	`,
+	}
+	cmd.AddCommand(offlineUpdateCmd)
+	offlineUpdateCmd.Flags().BoolVarP(&prod, "prod", "", false,
+		"Instruct to fetch content of production Target")
+	offlineUpdateCmd.Flags().StringVarP(&tag, "tag", "", "",
+		"Target tag")
+}
+
+func doOfflineUpdate(cmd *cobra.Command, args []string) {
+	factory = viper.GetString("factory")
+	targetName := args[0]
+	dstDir := args[1]
+
+	if len(tag) == 0 {
+		subcommands.DieNotNil(errors.New("missing mandatory flag `--tag`"))
+	}
+
+	fmt.Printf("Checking whether Target exists; target: %s, tag: %s, production: %v\n", targetName, tag, prod)
+	ok, err := checkIfTargetExists(targetName)
+	subcommands.DieNotNil(err, "Failed to check whether Target exists:")
+	if !ok {
+		subcommands.DieNotNil(fmt.Errorf("The specified Target has not been found; target: %s, tag: %s, production: %v\n", targetName, tag, prod))
+	}
```

Comment:
I still don't get why do you need the `ok, err := checkIfTargetExists(targetName)` above.
You might just `return fmt.Errorf("The specified Target has not been found ...")` where you currently `return false, nil` from inside `checkIfTargetExists` to make the code flow simpler.

Reply:
looks like it should work.

Reply:
Actually it does not work exactly as I would like it, no way to distinguish between error case and non error case when targets are not found.
IMO, it is not correct to return an error if the check was successful but a specified target was not found.

Reply:
Hmm... I don't quite understand what the problem is?

IIUC from the perspective of _this script_ no found target is an error because it cannot continue, isn't it?
If yes, then there is no reason to distinguish between _this error_ versus _other error_, other than a bit different error message: in any case the program should terminate with the same exit code.

This is scripting application, so I don't buy purely philosophical categories like exception versus error; lets stay practical.